path_role.go

```diff
@@ -0,0 +1,433 @@
+	defer b.roleMutex.Unlock()
+
+	return b.nonLockedSetOCIRole(ctx, s, roleName, roleEntry)
+}
+
+// nonLockedSetOCIRole creates or updates a role in the storage. This method
+// does not acquire the write lock before reading the role from the storage. If
+// locking is desired, use lockedSetOCIRole instead.
+func (b *backend) nonLockedSetOCIRole(ctx context.Context, s logical.Storage, roleName string,
+	roleEntry *OCIRoleEntry) error {
+	if roleName == "" {
+		return fmt.Errorf("missing role name")
+	}
+
+	if roleEntry == nil {
+		return fmt.Errorf("nil role entry")
+	}
+
+	entry, err := logical.StorageEntryJSON("role/"+strings.ToLower(roleName), roleEntry)
+	if err != nil {
+		return err
+	}
+
+	if err := s.Put(ctx, entry); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// nonLockedOCIRole returns the properties set on the given role. This method
+// does not acquire the read lock before reading the role from the storage. If
+// locking is desired, use lockedOCIRole instead.
+// This method also does NOT check to see if a role upgrade is required. It is
+// the responsibility of the caller to check if a role upgrade is required and,
+// if so, to upgrade the role
+func (b *backend) nonLockedOCIRole(ctx context.Context, s logical.Storage, roleName string) (*OCIRoleEntry, error) {
+	if roleName == "" {
+		return nil, fmt.Errorf("missing role name")
+	}
+
+	entry, err := s.Get(ctx, "role/"+strings.ToLower(roleName))
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result OCIRoleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role"), nil
+	}
+
+	b.roleMutex.Lock()
+	defer b.roleMutex.Unlock()
+
+	return nil, req.Storage.Delete(ctx, "role/"+strings.ToLower(roleName))
+}
+
+func (b *backend) pathRoleList(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	b.roleMutex.RLock()
+	defer b.roleMutex.RUnlock()
+
+	roles, err := req.Storage.List(ctx, "role/")
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(roles), nil
+}
+
+func (b *backend) pathRoleRead(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleEntry, err := b.lockedOCIRole(ctx, req.Storage, strings.ToLower(data.Get("role").(string)))
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, nil
+	}
+
+	return &logical.Response{
+		Data: roleEntry.ToResponseData(),
+	}, nil
+}
+
+// create a Role
+func (b *backend) pathRoleCreate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	roleName := strings.ToLower(data.Get("role").(string))
+	if roleName == "" {
```

Comment:
Not required. This is a path parameter, so if not present it will be caught by the framework.